### ISSUE TITLE
SPIRE agent cannot verify identity within own pod

### ISSUE THREAD

**Issue by critterjohnson** [spiffe/spire#4917]
* **Version**: `1.8.5` / `latest`
* **Platform**: `linux/amd64`
* **Subsystem**: agent

The SPIRE agent cannot verify the identity of other processes running in the same pod. This is because of the `cgroup` for the PID the agent can see for intra-pod containers does not contain the PodUID.

## Steps to reproduce

I discovered this while trying to secure the `spire-agent` telemetry endpoint behind an Envoy deployed as a sidecar in the same pod, using the agent for SDS. However, there is a much easier way to reproduce this:

1. Follow the [Quickstart for Kubernetes](https://spiffe.io/docs/latest/try/getting-started-k8s/) example up to "Configure and deploy the SPIRE Agent".

2. Replace `agent-daemonset.yaml` with the following and apply:

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: spire-agent
  namespace: spire
  labels:
    app: spire-agent
spec:
  selector:
    matchLabels:
      app: spire-agent
  template:
    metadata:
      namespace: spire
      labels:
        app: spire-agent
    spec:
      hostPID: true
      hostNetwork: true
      dnsPolicy: ClusterFirstWithHostNet
      serviceAccountName: spire-agent
      initContainers:
        - name: init
          # This is a small image with wait-for-it, choose whatever image
          # you prefer that waits for a service to be up. This image is built
          # from https://github.com/lqhl/wait-for-it
          image: cgr.dev/chainguard/wait-for-it
          args: ["-t", "30", "spire-server:8081"]
      containers:
        - name: client
          image: ghcr.io/spiffe/spire-agent:1.5.1
          command: ["/opt/spire/bin/spire-agent"]
          args: [ "api", "watch",  "-socketPath", "/run/spire/sockets/agent.sock" ]
          volumeMounts:
            - name: spire-agent-socket
              mountPath: /run/spire/sockets
              readOnly: true
        - name: spire-agent
          #image: ghcr.io/spiffe/spire-agent:1.5.1
          image: critterjohnson/spire-agent:latest
          args: ["-config", "/run/spire/config/agent.conf"]
          env:
          - name: MY_NODE_NAME
            valueFrom:
              fieldRef:
                fieldPath: status.podIP
          volumeMounts:
            - name: spire-config
              mountPath: /run/spire/config
              readOnly: true
            - name: spire-bundle
              mountPath: /run/spire/bundle
            - name: spire-agent-socket
              mountPath: /run/spire/sockets
              readOnly: false
          livenessProbe:
            httpGet:
              path: /live
              port: 8080
            failureThreshold: 2
            initialDelaySeconds: 15
            periodSeconds: 60
            timeoutSeconds: 3
          readinessProbe:
            httpGet:
              path: /ready
              port: 8080
            initialDelaySeconds: 5
            periodSeconds: 5
      volumes:
        - name: spire-config
          configMap:
            name: spire-agent
        - name: spire-bundle
          configMap:
            name: spire-bundle
        - name: spire-agent-socket
          hostPath:
            path: /run/spire/sockets
            type: DirectoryOrCreate
```

The only difference this has from the regular Daemonset is that the agent container from the `client` Deployment was copied in.

3. Register the node as normal, but instead of the provided command for the workload, run

```
kubectl exec -n spire spire-server-0 -- \
    /opt/spire/bin/spire-server entry create \
    -spiffeID spiffe://example.org/ns/default/sa/default \
    -parentID spiffe://example.org/ns/spire/sa/spire-agent \
    -selector k8s:ns:spire \
    -selector k8s:sa:spire-agent
```

4. Check the `client` container logs: `kubectl logs -n spire spire-agent-4tnlr -c client` (replacing `4tnlr` with the uid given to the pod). You'll see a ton of `no identity issued` logs (which also appear in the agent, of course).

To confirm that what I listed is indeed the issue, I forked the repo and added a regex for the `cgroup` without pod UID (which the logic was already there to support, just not the correct regex prefix of `cri-`). The commented out line in the above Daemonset that changes the agent image points to my forked build. If you re-apply the manifests with that image, the client will receive its SVID.

**Comment by nikotih** [comment]
Are you running on cgroups v2 and cgroups namespace private by any chance? 

https://kubernetes.io/docs/concepts/architecture/cgroups/#check-cgroup-version

If yes, see https://github.com/spiffe/spire/issues/4004

**Comment by critterjohnson** [comment]
> Are you running on cgroups v2 and cgroups namespace private by any chance?
> 
> https://kubernetes.io/docs/concepts/architecture/cgroups/#check-cgroup-version
> 
> If yes, see #4004

@nikotih yep, that is 100% the issue - didn't see that issue when I was checking if this was already reported. Either way, I think the band-aid solution until cgroups v2 is fully supported is to include a cgroup RE that does not have pod UID. I've got a fork I'm hacking on now that works, though I need to clean it up and get it passing tests (the regex is too loose right now).

It won't fix the fact that the agent cannot get its own containerID, but it can get the containerID of other procs in the same pod.

**Comment by azdagron** [comment]
Yes, happy to.

**Comment by critterjohnson** [comment]
@azdagron FWIW, I have a working patch that fixes this issue (not the overall cgroups issue) by just updating the regex. It works for our purposes and I'm happy to PR upstream as a quick fix to this issue while you guys work on #4004.
